### PR TITLE
[5.6] Execute an Artisan command using either its name or class

### PR DESCRIPTION
**Pull request by KennedyTedesco**
This implementation adds the ability to execute a command using class notation:

```php
Artisan::call(FooCommand::class, [
    'user' => 1, '--queue' => 'default'
]);
```

Using the command name:

```php
Artisan::call('foo:bar', [
    'user' => 1, '--queue' => 'default'
]);
```

### CODE REVIEW

**Comment by m1guelpf** [comment]
@KennedyTedesco Maybe you should check if the class is a command and if it's registered...

**Comment by KennedyTedesco** [comment]
@m1guelpf Hi, there! Thanks for your comment. :) I don't think it's necessary. It's a developer responsibility pass a valid command class or name, otherwise an error will be throw, anyway.

**Comment by laurencei** [comment]
So this removes the requirement to register commands with the console? You could run *anything* from there?

What is the benefit? We already have auto-loading command logic that @themsaid did in 5.5 - why would we ever want to call the class directly?

**Comment by KennedyTedesco** [comment]
Actually, it's not about auto-loading or anything. All those stuff remains the same. This implementation's just an easy-way to call a command using its class. It just instantiates the class to extract the command name. 😊 

We already have this feature when scheduling a task:

https://laravel.com/docs/5.6/scheduling#scheduling-artisan-commands

```php
$schedule->command(EmailsCommand::class, ['--force'])->daily();
```

**Comment by arjasco** [comment]
Yeah, I’m for this. I personally would change it to.

‘’’php
if ($command instanceof Command) {
    $command = $this->laravel->make($command)->getName();
}
‘’’
That way it’s definitely a Command subclass no need to check it it exists because it would if it’s an instance of it anyway..

**Comment by KennedyTedesco** [comment]
Using `class_exists()` is the simplest way because if the developer inform a class, it need to be a command, otherwise, an error will be throw (checking it or not).

Changing the approach as you suggest, we would need to:

- Check if $command is a class;
- Check if $command is an instance of Command;
- Finally, get its name.

I'm using this approach:

https://github.com/laravel/framework/blob/5.6/src/Illuminate/Console/Scheduling/Schedule.php#L77

**Comment by arjasco** [comment]
Well if it’s an instance of Command it’s a class.. so the first check doesn’t need to happen.

I agree both being kept consistent if scheduling already functions that way. Just feels weird being able to give both of them a random class that has the method getName and it will work if the command name in question exists..

I would think making sure it’s actually a Command instance just seems sensible otherwise as said above some completely different instance could provide a valid command name which does work.. but syntactically doesn’t make sense.

**Comment by KennedyTedesco** [comment]
`$command instanceof Command` it's not a valid statement because `EmailsCommand::class` is not a instance (yet).

So, in order to check this way, at the minimum we would need to:

```php
if (class_exists($command) && resolve($command) instanceof Command) {
   // TODO
}
```

**Comment by arjasco** [comment]
Ahh! Apologies, I missed that 🙈this is most likely best route then.

Unless ‘is_subclass_of’ works for this case?

http://php.net/manual/en/function.is-subclass-of.php

**Comment by KennedyTedesco** [comment]
@arjasco This definitely do the job elegantly. Thank you so much.

If this PR gets merged i'll change that part on the Schedule too.

File: tests/Integration/Console/ConsoleApplicationTest.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+        $this->assertEquals($exitCode, 0);
+    }
+
+    /**
+     * @expectedException \Symfony\Component\Console\Exception\CommandNotFoundException
+     */
+    public function test_artisan_call_invalid_command_name()
+    {
+        $this->artisan('foo:bars');
```
Comment (taylorotwell) [inline tests/Integration/Console/ConsoleApplicationTest.php]: @KennedyTedesco this test freezes entirely on master branch. Can you fix it?

Comment (KennedyTedesco) [reply tests/Integration/Console/ConsoleApplicationTest.php]: @taylorotwell We can fix on our side by evaluating if the command really exists. What you think about it?

https://github.com/KennedyTedesco/framework/commit/bc20aaa0ef07682cf0ef7d4c9eb1b11456c0f07e

I can send a pull request if it's fine.

Comment (taylorotwell) [reply tests/Integration/Console/ConsoleApplicationTest.php]: OK

**Pull request merged**